### ISSUE TITLE
some names not recognized as varaiables

### ISSUE THREAD

**Issue by hax** [pug-php/pug#54]
``` pug
p= a + b
```

expect compiled to `$a + $b`,
actual `$a + b`, which cause runtime exception `Notice: Use of undefined constant b`

It seems all binary operators (+, -, *, /, %, &&, ||, etc.) have same problem.

**Comment by hax** [comment]
Ooops, unary operators (!, -, ~) also have the same problem.

**Comment by kylekatarnls** [comment]
You expect `+` compile to `$a + $b`, but some people who take templates from a JS environment would expect `$a . $b`. The operators remain a big problem and touch them is a big risk of regressions for some users. This example is simple but imagine:

``` pug
p(a=(a+=b + c(d, '+', e + f))= g(h + i) + j("= a + b", a + b)
```

That can become very tricky. We detect variables and append missing `$` after `,` and `(`, but this is just to help to migrate from pug-js.

If you want the exact native pug-js behaviour, you can use a wrapper, but you won't be able to call PHP functions/methods or to use instance of your classes in the templates.

That's why I do not plan right now to touch operators and I encourage you to explicitly write PHP code:

``` pug
p= $a + $b
```

That's pug-php, the goal is to be able use PHP stuff in pug templates, so this is not a priority.

Thanks anyway.

**Comment by hax** [comment]
Well, I think wether `+` should be treat as number add operator or string concat operator is a separate issue. Let me use `*` as example. What I argue here is `a * b` be compiled to `$a * b` is weird and error-prone (it's runtime error and very hard to locate the source in complex case).

From the old jade-php there are many effort to allow both js-like expression or php-like expression. But I had to say it's an unachievable goal and introduce many issues. I totally agree that just write php expression (eg. `$a * $b`) is a more clear and stable way in pug-php.

In fact the code style of my team require writing php expression in jade-php. But we still meet many syntax problems because there is no option which just stick on php expression and report all syntax error (with template line/col info).

The ideal way I can imagine is:

Add a new option `expressionLanguage` which could be `php`, `js-like`, `you-guess`.

If `php`, `a * b` is compiled to `a * b` (just treat them as const, though we'd better warn the user for it's not follow UPPER_CASE const name convension)

If `js-like`, `a * b` is compiled to `$a * $b`

If `you-guess`, `a * b` is compiled to `$a * b` 😁

**Comment by kylekatarnls** [comment]
It's a good idea, I like this approach and I will study possibilities. If you want to submit a pull-request for this, please let me know.

**Comment by kylekatarnls** [comment]
I started to get my hands dirty with expressions handle, and it need to be entirely refactored.

Also, such JS to PHP syntax conversion are yet used by different template engines and framework and is not specific to Pug. So I will try to implement an existing library if I find a good one. Else I will create a separate engine.

**Comment by kylekatarnls** [comment]
Hi, now on the branch js-phpize (use dev-js-phpize version in composer) with the option `'expressionLanguage' => 'js'`, you can handle some basic JS syntaxes:

``` pug
p=a + b
p=a * (b - foo.bar)
```

For the moment, it's available in attributes and tag contents, it wil be introduced step by step in other locations.

And this first step will be merged with the next version.

Thanks,

**Comment by kylekatarnls** [comment]
Available now in Version 2.4.0.

**Comment by kylekatarnls** [comment]
Complex calculs with js variable syntaxes are now possible in expressionLanguage js mode.
